presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java

```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
     private void extractFromSource(SearchHit hit)
     {
         Map<String, Object> map = hit.getSourceAsMap();
+        TreeMap<String, Object> fieldsMap = new TreeMap<>(new Comparator<String>()
+        {
+            @Override
+            public int compare(String key1, String key2)
+            {
+                int length1 = key1.split("\\.").length;
+                int length2 = key2.split("\\.").length;
+                if (length1 == length2) {
+                    return key1.compareTo(key2);
+                }
+                return length2 - length1;
+            }
+        });
         for (Map.Entry<String, Object> entry : map.entrySet()) {
-            String jsonPath = entry.getKey();
-            Object entryValue = entry.getValue();
-            setFieldIfExists(jsonPath, entryValue);
+            fieldsMap.put(entry.getKey(), entry.getValue());
+        }
+        processFields(fieldsMap);
+    }
+
+    private void processFields(TreeMap<String, Object> fieldsMap)
+    {
+        if (fieldsMap.size() == 0) {
+            return;
+        }
+        Map.Entry<String, Object> first = fieldsMap.firstEntry();
+        String field = first.getKey();
+        if (field.contains(".")) {
+            String prefix = field.substring(0, field.lastIndexOf('.'));
+            Map<String, Object> value = new HashMap<>();
+            int size = field.split("\\.").length;
+            Iterator<String> iterator = fieldsMap.navigableKeySet().iterator();
+            while (iterator.hasNext()) {
+                String name = iterator.next();
+                if (name.split("\\.").length == size && name.startsWith(prefix)) {
+                    value.put(name.substring(name.lastIndexOf('.') + 1), fieldsMap.get(name));
+                    iterator.remove();
+                    continue;
+                }
+                break;
```

Comment:
if we break for the first iteration, we don't need a while loop

Reply:
need to loop processing all nested json paths with same component length, e.g.
a.b.c.d and a.b.c.f
continue and while are cooperating to make sure json paths with the biggest component length and with the same prefix are processed in each while loop